src/Symfony/Component/String/GraphemeString.php

```diff
@@ -0,0 +1,346 @@
+            $needle = (string) $needle;
+        }
+
+        $form = null === $this->ignoreCase ? \Normalizer::NFD : \Normalizer::NFC;
+        normalizer_is_normalized($needle, $form) ?: $needle = normalizer_normalize($needle, $form);
+
+        if ('' === $needle || false === $needle) {
+            return null;
+        }
+
+        $string = $this->string;
+
+        if (0 > $offset) {
+            // workaround https://bugs.php.net/74264
+            if (0 > $offset += grapheme_strlen($needle)) {
+                $string = grapheme_substr($string, 0, $offset);
+            }
+            $offset = 0;
+        }
+
+        $i = $this->ignoreCase ? grapheme_strripos($string, $needle, $offset) : grapheme_strrpos($string, $needle, $offset);
+
+        return false === $i ? null : $i;
+    }
+
+    public function join(array $strings): AbstractString
+    {
+        $str = parent::join($strings);
+        normalizer_is_normalized($str->string) ?: $str->string = normalizer_normalize($str->string);
+
+        return $str;
+    }
+
+    public function length(): int
+    {
+        return grapheme_strlen($this->string);
+    }
+
+    /**
+     * @return static
+     */
+    public function normalize(int $form = self::NFC): parent
+    {
+        $str = clone $this;
+
+        if (\in_array($form, [self::NFC, self::NFKC], true)) {
+            normalizer_is_normalized($str->string, $form) ?: $str->string = normalizer_normalize($str->string, $form);
+        } elseif (!\in_array($form, [self::NFD, self::NFKD], true)) {
+            throw new InvalidArgumentException('Unsupported normalization form.');
+        } elseif (!normalizer_is_normalized($str->string, $form)) {
+            $str->string = normalizer_normalize($str->string, $form);
+            $str->ignoreCase = null;
+        }
+
+        return $str;
+    }
+
+    public function prepend(string ...$prefix): AbstractString
+    {
+        $str = clone $this;
+        $str->string = (1 >= \count($prefix) ? ($prefix[0] ?? '') : implode('', $prefix)).$this->string;
+        normalizer_is_normalized($str->string) ?: $str->string = normalizer_normalize($str->string);
+
+        if (false === $str->string) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    public function replace(string $from, string $to): AbstractString
+    {
+        $str = clone $this;
+        normalizer_is_normalized($from) ?: $from = normalizer_normalize($from);
+
+        if ('' !== $from && false !== $from) {
+            $tail = $str->string;
+            $result = '';
+            $indexOf = $this->ignoreCase ? 'grapheme_stripos' : 'grapheme_strpos';
+
+            while (false !== $i = $indexOf($tail, $from)) {
+                $slice = grapheme_substr($tail, 0, $i);
+                $result .= $slice.$to;
+                $tail = substr($tail, \strlen($slice) + \strlen($from));
+            }
+
+            $str->string = $result .= $tail;
```

Comment:
```suggestion
            $str->string = $result . $tail;
```
Reassignation doesn't seem to be necessary here

Reply:
It's more memory-efficient as it doesn't need to keep a partial string (`$result`) in memory.